Do complete Refactoring of the UI

RaHoni opened the pull request:
This includes 
- splitting the commandhandler in destinctive functions for easier return value handling
- changing the ipc socket to send serialized class instances
- adding a cli to the console

RaHoni committed:
data: Made data Package private and fixed a delete error

When Deleting a switch we would previously try to delete the switch first this would fail because of the limit for mhz and mqtt.

RaHoni committed:
uiControl: deprecated receiveCommand

Deprecated receiveCommand and replaced it with individual methods for each command.

RaHoni committed:
repl + websocket: Changed to independent answers

Changed websocket and stringCommandHandler to use the new independent methode api.

While doing this also fix some Protocol issues at the Client

RaHoni committed:
Made mqttController.keyArray Private for package PrivateHome as so that getting the key is harder.

RaHoni committed:
Replaced switchType and controlType strings with enums

This makes storing it in the database more efficient and also makes validity checking easier.

RaHoni committed:
Changed the communication protocol of the socket to java objects

RaHoni committed:
Extendet the .gitignore

RaHoni committed:
Made read response typed so that you don't need to cast it every time.

RaHoni committed:
Fused ipcSuccessResponse and ipcErrorResponse in ipcSuccessResponse

ipcSuccessResponse now contains the original Command the success state
and an optional throwable. This was needed for the new read methode in
the console.

RaHoni committed:
Upgrading project dependecys

RaHoni committed:
Added sbtBuildInfo

This can be used to make sbt variables available in Scala this is used for the upcomming cli version output.
It can also used to generate the licens notices in the future.

RaHoni committed:
Added a basic cli

With this version of the cli you can display a help dialog the current
version that gets read from sbt. As real functions you can display the
current status of a switch and also turn one on, off and toggle one with
a specific ID.
If you specify -i and don't specify a valid used ID it will launch the
interactive ID selector.

RaHoni committed:
Fixup: reenabling sbt-buildinfo

RaHoni committed:
cliParser: changed the id validation trait to a class (subcommandWithId)

This has the advantage that all classes that inherit from
subcommandWithId will inherit id as option and also verify it. This is
especially useful when adding checks.

RaHoni committed:
Added scalatest for cliParser.scala

RaHoni committed:
Made the ObjectOutputStreams non Buffered

This removes the need to flush the streams which always only lead to
errors for us if really needed they can be added again later

maximiliani committed:
Update README.md

maximiliani commented on src/main/scala/PrivateHome/UI/repl.scala:
```diff
@@ -25,52 +25,64 @@ import org.slf4j.LoggerFactory
-
 object repl {
-  private val logger = LoggerFactory.getLogger(this.getClass)
   val socketPath = "/tmp/privatehome2.sock"
```
Why is the socket named "privatehome2.sock"?

RaHoni replied:
Because of the old socketfile which was simply privatehome.sock

RaHoni replied:
This is fixed now

maximiliani commented on src/main/scala/PrivateHome/data.scala:
```diff
@@ -223,7 +225,7 @@ object data {
     }.map(rs => device(rs)).list().apply().foreach(d => {
       d.switchtype match {
-        case "mqtt" =>
+        case Mqtt =>
```
All uppercase would look nicer, as it is an own type, right?

RaHoni replied:
Yea

maximiliani reviewed (dismissed):
Good job.  For the next time please write some more comments and do some more testing.

RaHoni commented:
This isn't ready yet because there are missing test and much more

RaHoni committed:
Made the Enums controlType and switchType values CAPS

RaHoni committed:
Fixed a logic error in the validation of the id

RaHoni committed:
Changed the socket file for IPC to be the name of BuildInfo(sbt project)

RaHoni committed:
Removed an unused match case and replaced it with an asInstanceOf call.

RaHoni committed:
Moved the error check of IOException to the end because it is the super
class of all the other checks and would thereby always trigger.

maximiliani committed:
Update issue templates

maximiliani committed:
Create CONTRIBUTING.md

maximiliani committed:
Rename CONTRIBUTING.md to CONTRIBUTING

maximiliani committed:
Create CODE_OF_CONDUCT.md

RaHoni committed:
Setting: Moved the settingsPath to separate object.

This is needed so that it can be changed for testing. Because
otherwise the file gets loaded before the path can be changed.

RaHoni committed:
Upgraded akka

RaHoni committed:
IPC: Added the response type to the IPCCommands this is used to have a defined returntype for all Commands.

RaHoni committed:
Removed parentese form pin request.

RaHoni committed:
Added missing parantheses.

RaHoni committed:
Deprection: Removed editXml thirby removing all deprections

RaHoni committed:
IPC: Added an IPCPingResponse for the new response Type

RaHoni committed:
PrivateHome: Fixing a missing parenthese that was marked as deprecation.

RaHoni committed:
console: Changed send to also receive the Answer

Because now that every Command has the Type of the Answer assigned in it, we can use this to receive an answer without the need to specify the Answer type and can receive it automatically.

RaHoni committed:
console: Added adController Command

RaHoni committed:
console: Added a function for Password readin

This modulates the reading of passwords so that it is easier to use in different places.

RaHoni committed:
console: Added getControllerId

This is used to get the ID of a controller and verify that it is a valid one.

RaHoni committed:
console: Optimized getDeviceID

Moved an if check up so that it only needs to be done ones and also to have the possibility to skip more.

RaHoni committed:
console: reorganised Code

Reorder code

RaHoni committed:
console: Added Tests with above of 80% coverage only main and help aren't tested

RaHoni committed:
console: Added Scaladoc to all functions

RaHoni committed:
Update README.md

RaHoni committed:
cliParser: Fix a logical error

The maximum Value for percentage for on should be 100 not 99

RaHoni committed:
Add cleanKeepFiles

This will prevent the sbt clean command from deleting the scoverage
report.

RaHoni committed:
IPC: Added parameter checks for ipc[Off/On]Command

RaHoni committed:
Console: Adding a bunch of functions to console

Added functions include addController, ensureConnection, a
readPassword helper

Additionally send will also directly return the answer from the Server.
Added parameters to a bunch of functions to define defaults to use by
the cli.

RaHoni committed:
console: Started with writing Tests

RaHoni committed:
Readded the send cast fail detection and add Switch

These were lost while merging

RaHoni committed:
Added a Helper for percent in UI.on

Now there is a helper that will always return the specified percentage as a float and wenn nothing is spezified it will return 1.0f therefore it was possible to make the specific options private.

RaHoni committed:
Moved the execution of the subcimmands to an helper

RaHoni committed:
Fixed a typo in the validity check for the IDs

RaHoni committed:
Added an exit command to the console.

maximiliani committed:
Update CONTRIBUTING

RaHoni committed:
Merge branch 'dev' into uiControl_Refactoring

maximiliani committed:
Update CONTRIBUTING

RaHoni commented:
This is far enough to be merged and extended in the future
